WordPress/openverse
# Enable running the snapshot tests for both color modes

## Fixes


Extracted from #4915
Related to #4305 by @zackkrida

## Description


This is one idea of how we can implement testing both dark and light theme for visual regressions **without running each test twice**. It relies on the `expectSnapshot` function, which now runs 2 snapshot comparisons instead of one. 

First, the default snapshot is taken. Then, the theme switcher is used to change the color mode, and the second snapshot is taken. To make sure that we **always run both snapshot checks**, even if the first one fails, we use the `expect.soft` that runs the test to the end, but fails if one of the `expect`s in the test fails.

We use the theme switcher to change the color mode. To do this, we add a `themeSwitcher` to all stories using a decorator, and use the footer switcher for the tests that use the Nuxt app.

`expectSnapshot` function is extracted from the `breakpoints.ts` file. It needs the `page` argument from Playwright test to allow locating the theme switcher. The optional (nullable) parameters were combined to a separate `options` object, which allows passing only the necessary options instead of having to pass `{}` blank object for the unused one. 

This PR also adds the theme switcher to the Storybook toolbar, so now you can view both color modes. The Storybook switcher's state is synced with the state of our own theme switcher in the story.

The image cell snapshots had to be updated. I made the width smaller to match the real size of images in the grid, and to make the snapshot files smaller, and to fit all of them into a single screen height. [View updated image cell story]( https://docs.openverse.org/_preview/4926/storybook/?path=/story/components-vimagecell--default&globals=viewport:xs)

## Testing Instructions

Go to [Storybook](https://docs.openverse.org/_preview/4926/storybook/?path=/story/components-vbutton--default) and see check that the color theme switcher is working. Also, check the follow-up PRs that add the dark snapshots.

## Checklist


      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
